docs/language/builtins.rst

```diff
@@ -311,6 +311,35 @@ Its underlying type is ``bytes32``, but it will be reported correctly as the ``H
 .. include:: ../examples/substrate/hash_type.sol
   :code: solidity
 
+chain_extension(uint32 ID, bytes input) returns (uint32, bytes)
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Only available on Substrate. Call the chain extension with the given ``ID`` and ``input`` data.
+Returns the return value from the chain extension and the output data.
+
+This function is a low level interface.
+The caller is responsible for encoding the input and decoding the output correctly.
+We expect parachain authors to write their own higher level libraries on top.
+
+.. warning::
+    This function calls the runtime API `call_chain_extension <https://docs.rs/pallet-contracts/latest/pallet_contracts/api_doc/trait.Version0.html#tymethod.call_chain_extension>`.
+	It assumes that the implementation of the chain extension does
+	- read the input from the ``input_ptr`` parameter, used as a buffer pointer
+	- write potential output into the buffer found at the ``output_ptr`` pointer
+	- respect the output buffer length in ``output_len_ptr`` to prevent OOB writes. The output buffer is 16KB in size.
+	- write the amount of bytes written to ``output_ptr`` into the buffer at ``output_len_ptr``
```

Comment:
```suggestion
	It assumes that the implementation of the chain extension
	- reads the input from the ``input_ptr`` parameter, used as a buffer pointer
	- writes potential output into the buffer found at the ``output_ptr`` pointer
	- respects the output buffer length in ``output_len_ptr`` to prevent OOB writes. The output buffer is 16KB in size.
	- writes the amount of bytes written to ``output_ptr`` into the buffer at ``output_len_ptr``
```